### DISCUSSION TITLE
make `MachineConfig` type work with Typegen

### DISCUSSION THREAD

**Discussion by amir-arad** [statelyai/xstate#3530] [Ideas]
currently If I want to define a machine config object outside of the `createMachine(...)` function **and** use Typegen, I have to define all generic parameters like so: 
```typescript 
const statusConfig: MachineConfig<
    StatusContext,
    any,
    { type: StatusEvent },
    BaseActionObject,
    ServiceMap,
    TypegenEnabled
> = {...}
```
since `TTypesMeta` is the last of the generic arguments and defaults to `TypegenDisabled`.
Obviously there is a TS inference issue there (it should detect that I'm supplying `tsTypes` and set to `TypegenEnabled`). From my experience with TS, it's not super great with complex inference scenarios, and might even change behavior within the same major version when the typing voodo is deep enough.
how about having a type alternative to `MachineConfig`, which accepts the `typegen` type and infers the rest from it? something like `MachineConfigTypegen<import('./driver.typegen').Typegen0>` ? this will of course mean adding all the types to typegen (context, events, etc.)

**Comment by Andarist** [comment]
> If I want to define a machine config object outside of the createMachine(...)

It would be very much appreciated if you could motivate this desire.
